clippy_lints/src/methods/join_absolute_path.rs

```diff
@@ -0,0 +1,30 @@
+use clippy_utils::{diagnostics::span_lint_and_sugg, ty::is_type_diagnostic_item};
+use rustc_ast::ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::LateContext;
+use rustc_span::{symbol::sym::Path, Span};
+
+use super::JOIN_ABSOLUTE_PATH;
+
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>, join_arg: &'tcx Expr<'tcx>, span: Span) {
+    let ty = cx.typeck_results().expr_ty(expr);
```

Comment:
With the updated receiver, this gives you the following types (using `eprintln!("ty:#?")`)

```
&std::path::Path
&std::path::Path
&[&str]
&std::path::Path
```

The reference at the start prevents the `if is_type_diagnostic_item(cx, ty, Path);` check from passing. Instead, we should first remove the outer reference, this can be done with [`Ty::peel_refs`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Ty.html#method.peel_refs)

```suggestion
    let ty = cx.typeck_results().expr_ty(expr).peel_refs();
```